Support generic dynamic shape with padding

nandesuka opened the pull request:
Summary:
Inductor has the following configurations:

config.comprehensive_padding
config.padding_alignment_bytes
config.padding_stride_threshold

In the case of static shape by enabling these three options Inductor will generate code for Flexible layout tensors that tries to pad up all stride dimension to be a multiple of config.padding_alignment_bytes for strides above: config.padding_stride_threshold. In the case where dynamic shapes is enabled no padding is done today.
This PR introduces the following configuration which allows the user to specify they wish to generated a padded stride even in the case of dynamic shape operations. This is mainly done so we don't break the previous behaviour of not padding up dynamic shape use cases. The config.padding_stride_threshold does not apply since the values of the strides are dynamic.

config.pad_dynamic_shapes

In addition to this a new mode "python_slow" has been added to launch grid calculation which achieves the same ceildiv behaviour that is generally applicable to integer division. This is done to prevent test regressions and make wrapper_fxir codegen more generic.

Test Plan:
CI

Rollback Plan:

Differential Revision: D80468808

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @mlazos

blaine-rister commented on torch/_inductor/codegen/wrapper_fxir.py:
```diff
@@ -460,8 +460,10 @@ def _generate_sym_node(
             def replace_floor_div(expr: sympy.Expr) -> sympy.Expr:
                 """
-                Converts floor(x / c) to x // c.
+                Converts -x / c or to (x + c - 1) / c
```
typo?

blaine-rister commented on torch/_inductor/config.py:
```diff
@@ -899,6 +899,9 @@ def decide_compile_threads() -> int:
 )
 pad_channels_last = False
 
+# Control if we will do padding on dynamic shapes
```
Is this to avoid regressions on existing tests? I guess another option would be to set `comprehensive_padding=False` on those tests, but that could be a more invasive change. I'll defer to the other reviewers on this one.

blaine-rister reviewed (commented):
Nice PR! This mostly LGTM. I left a few nits and a question about 3d testing.

nandesuka commented on torch/_inductor/ir.py:
```diff
@@ -3764,11 +3755,17 @@ def _pad_strides(
+            ) or (isinstance(stride, sympy.Expr) and config.pad_dynamic_shapes)
+            new_strides[idx] = stride
+            if require_padding:
+                new_strides[idx] = ceildiv(stride, align) * align
```
It looks something like this:

(8*s48*(((s87 + 7)//8)), 8*(((s87 + 7)//8)), 1)

blaine-rister reviewed (approved):
LGTM! Very nice PR.

nandesuka committed:
Support generic dynamic shape with padding (#160997)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/160997

Inductor has the following configurations:

config.comprehensive_padding
config.padding_alignment_bytes
config.padding_stride_threshold

In the case of static shape by enabling these three options Inductor will generate code for Flexible layout tensors that tries to pad up all stride dimension to be a multiple of config.padding_alignment_bytes for strides above: config.padding_stride_threshold. In the case where dynamic shapes is enabled no padding is done today.

This PR introduces the following configuration which allows the user to specify they wish to generated a padded stride even in the case of dynamic shape operations. This is mainly done so we don't break the previous behaviour of not padding up dynamic shape use cases. The config.padding_stride_threshold does not apply since the values of the strides are dynamic.

config.pad_dynamic_shapes

In addition to this a new mode "python_slow" has been added to launch grid calculation which achieves the same ceildiv behaviour that is generally applicable to integer division. This is done to prevent test regressions and make wrapper_fxir codegen more generic.

Test Plan:
CI

Rollback Plan:

Reviewed By: blaine-rister

Differential Revision: D80468808
